File: src/clients/metadata.js
/**
 * Takes a list of ids and sends a request to the metadata
 * service to filter those IDs to just those with the corresponding AVUs and
 * target types
 */

import * as config from "../configuration";
import fetch from "node-fetch";

/**
 * @param username - The username of the authenticated user
 * @param targetTypes - Array of the types of resources to include
 * @param avus - Array of the metadata avus to use for filtering
 * @param targetIds - Array of the resource IDs to include and filter
 * @returns {Promise<*>}
 */
export const getFilteredTargetIds = async ({
    username,
    targetTypes,
    avus,
    targetIds,
}) => {
    const reqURL = new URL(config.metadataURL);
    reqURL.pathname = "/avus/filter-targets";
    reqURL.search = new URLSearchParams({ user: username }).toString();
    const body = {
        "target-types": targetTypes,
        "target-ids": targetIds,
        avus,
    };

    const resp = await fetch(reqURL, {
        method: "post",
        body: JSON.stringify(body),
        headers: { "Content-Type": "application/json" },
    });

    if (!resp.ok) {
        const msg = await resp.text();
        throw new Error(msg);
    }

    const data = await resp.json();
    return data["target-ids"];
};

File: apis/metadata.go
package apis

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
)

type MetadataAPI struct {
	metadataURL *url.URL
}

func NewMetadataAPI(metadataURL *url.URL) *MetadataAPI {
	return &MetadataAPI{
		metadataURL: metadataURL,
	}
}

type TargetIDs struct {
	TargetIDs []string `json:"target-ids"`
}

func (m *MetadataAPI) GetFilteredTargetIDs(ctx context.Context, username string, targetTypes []string, avus []map[string]string, targetIDs []string) ([]string, error) {
	u := fixUsername(username)

	fullURL := *m.metadataURL.JoinPath("avus", "filter-targets")
	q := fullURL.Query()
	q.Set("user", u)
	fullURL.RawQuery = q.Encode()

	body := map[string]interface{}{}
	body["target-types"] = targetTypes
	body["target-ids"] = targetIDs
	body["avus"] = avus

	b, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, fullURL.String(), bytes.NewReader(b))
	if err != nil {
		return nil, err
	}
	req.Header.Set("content-type", "application/json")

	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	rb, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode != http.StatusOK {
		msg := string(rb)
		return nil, fmt.Errorf("url %s; status code %d; msg %s", fullURL.String(), resp.StatusCode, msg)
	}

	var data TargetIDs

	if err = json.Unmarshal(rb, &data); err != nil {
		return nil, err
	}

	retval := data.TargetIDs
	return retval, nil
}
